src/java/org/apache/cassandra/index/sai/StorageAttachedIndexGroup.java

```diff
@@ -0,0 +1,241 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+
+import org.apache.cassandra.db.ColumnFamilyStore;
+import org.apache.cassandra.db.DecoratedKey;
+import org.apache.cassandra.db.RegularAndStaticColumns;
+import org.apache.cassandra.db.WriteContext;
+import org.apache.cassandra.db.filter.RowFilter;
+import org.apache.cassandra.db.lifecycle.LifecycleNewTracker;
+import org.apache.cassandra.db.lifecycle.Tracker;
+import org.apache.cassandra.db.memtable.Memtable;
+import org.apache.cassandra.db.rows.Row;
+import org.apache.cassandra.index.Index;
+import org.apache.cassandra.index.sai.metrics.TableQueryMetrics;
+import org.apache.cassandra.index.sai.metrics.TableStateMetrics;
+import org.apache.cassandra.index.sai.plan.StorageAttachedIndexQueryPlan;
+import org.apache.cassandra.index.transactions.IndexTransaction;
+import org.apache.cassandra.io.sstable.Component;
+import org.apache.cassandra.io.sstable.Descriptor;
+import org.apache.cassandra.io.sstable.format.SSTableFlushObserver;
+import org.apache.cassandra.notifications.INotification;
+import org.apache.cassandra.notifications.INotificationConsumer;
+import org.apache.cassandra.notifications.MemtableDiscardedNotification;
+import org.apache.cassandra.notifications.MemtableRenewedNotification;
+import org.apache.cassandra.schema.TableMetadata;
+
+/**
+ * Orchestrates building of storage-attached indices, and manages lifecycle of resources shared between them.
+ */
+@ThreadSafe
+public class StorageAttachedIndexGroup implements Index.Group, INotificationConsumer
+{
+    private final TableQueryMetrics queryMetrics;
+    private final TableStateMetrics stateMetrics;
+    private final Set<StorageAttachedIndex> indices = Sets.newConcurrentHashSet();
+    private final ColumnFamilyStore baseCfs;
+
+    StorageAttachedIndexGroup(ColumnFamilyStore baseCfs)
+    {
+        this.baseCfs = baseCfs;
+        this.queryMetrics = new TableQueryMetrics(baseCfs.metadata());
+        this.stateMetrics = new TableStateMetrics(baseCfs.metadata(), this);
+
+        Tracker tracker = baseCfs.getTracker();
+        tracker.subscribe(this);
+    }
+
+    @Nullable
+    public static StorageAttachedIndexGroup getIndexGroup(ColumnFamilyStore cfs)
+    {
+        return (StorageAttachedIndexGroup)cfs.indexManager.getIndexGroup(StorageAttachedIndexGroup.class);
```

Comment:
Nit: missed whitespace
```suggestion
        return (StorageAttachedIndexGroup) cfs.indexManager.getIndexGroup(StorageAttachedIndexGroup.class);
```